### PR TITLE
delegates more auth scenarios to kubernetes apiserver

### PR DESCRIPTION
**Pull request by prydonius**
If the provided bearer token is not a valid OIDC token, the proxy will
return an unauthorized error. This is an issue when trying to get an app
like the Kubernetes Dashboard to go through this proxy, as the Dashboard
will make calls to the API using its own Service Account token as well
as the logged-in user's OIDC token.

This patch changes the proxy's behaviour to be more of a first check,
passing invalid tokens and requests without tokens to the Kubernetes API
server for a second attempt. This then allows tokens, that might be
valid apiserver tokens, but not valid OIDC tokens, to be accepted (or
rejected) by the Kubernetes apiserver. This won't be able to handle
client cert authentication, but can handle other types of Authorization
headers (e.g. Basic auth).

See https://github.com/prydonius/dashboard-oidc-proxy also, which
walks through setting up this proxy with the Kubernetes Dashboard.

cc @migmartri

### CODE REVIEW

File: pkg/proxy/proxy.go [hunk 1]
```diff
@@ -111,11 +111,24 @@ func (p *Proxy) RoundTrip(req *http.Request) (*http.Response, error) {
 	info, ok, err := p.reqAuther.AuthenticateRequest(req)
 	if err != nil {
 		klog.Errorf("unable to authenticate the request due to an error: %v", err)
+
+		if !ok {
+			klog.Info("not a valid OIDC token, forward to Kubernetes API as this may be a valid API token")
+			return p.clientTransport.RoundTrip(req)
+		}
+
```
Comment (prydonius) [inline pkg/proxy/proxy.go]: I'm a little confused about this case, is it possible for `AuthenticateRequest` to return an error and `ok = true`?

**Comment by JoshVanL** [comment]
Hi @prydonius, thanks for the PR! :smile: 

We've had some discussions and am not sure about the current implementation. I think we want to be a bit more strict about what we pass through and limit it to token based auth, i.e. allowing Service Account bearer tokens. We could do this by adding more validation to whether the header exists and also contains a valid looking token before forwarding. I think it also makes sense to have this behavior enabled via an experimental flag that is disabled by default.

What do you think about this? Happy to help wherever :smiley: 

/ok-to-test
/assign @prydonius

File: pkg/proxy/proxy.go (hunk 1 above)
Comment (JoshVanL) [reply pkg/proxy/proxy.go]: I'm not aware of a scenario where `err != nil & ok`

**Comment by prydonius** [comment]
Hey @JoshVanL, sorry for the late response, I've had a lot going on!

My main thinking behind this implementation was to act a bit more like the Kubernetes apiserver. As far as I understand, the apiserver will evaluate different auth paths (i.e. so client cert auth works alongside bearer token auth).

This mostly comes from my viewpoint of this proxy, which is just to have an additional way to authenticate users using OIDC, but to not stand in the way of other auth strategies. This then allows us to avoid having to use different endpoints depending on the type of auth needed, it would be nice to just use one endpoint for everything.

That said, your team might have a bigger picture here where this proxy is more locked-down and specific to OIDC authentication. If that's the case, your suggestion to limit this to only service account tokens for the desired purpose and to put this behind an experimental flag is valid.

Let me know what you think, also happy to jump on a call to discuss more!

**Comment by nickperry** [comment]
I definitely need the option to only proxy requests with OIDC tokens and NOT proxy x509 / SA token requests. This would let us use kube-oidc-proxy to harden our clusters against use of leaked x509 certs and SA tokens outside of the cluster.

**Comment by JoshVanL** [comment]
Hi @nickperry,

Yes, I agree. These options would always be *off* by default. I'm ganna spend some time this week to see what this might look like and update some other areas.

**Pull request closed without merging**